metricflow/plan_conversion/node_processor.py

```diff
@@ -59,14 +61,111 @@ class MultiHopJoinCandidate:
     lineage: MultiHopJoinCandidateLineage
 
 
-@dataclass(frozen=True)
+class PredicatePushdownState(Enum):
+    """Enumeration of constraint states describing when predicate pushdown may operate in a dataflow plan.
+
+    This is necessary for holistic checks against the set of potentially enabled pushdown operations, because the
+    scenario where we only allow time range updates requires careful overriding of other pushdown properties.
+
+    Note: the time_range_only state is a backwards compatibility shim for use with conversion metrics while
+    we determine how best to support predicate pushdown for conversion metrics. It may have longer term utility,
+    but ideally we'd collapse this into a single enabled/disabled boolean property.
+    """
+
+    DISABLED = "disabled"
+    FULLY_ENABLED = "fully_enabled"
+    ENABLED_FOR_TIME_RANGE_ONLY = "time_range_only"
+
+
+@dataclasses.dataclass(frozen=True)
 class PredicatePushdownParameters:
-    """Container class for managing filter predicate pushdown.
+    """Container class for managing information about whether and how to do filter predicate pushdown.
 
-    Stores time constraint information for applying pre-join time filters.
+    The time_range_constraint property holds the time window for setting up a time range filter expression.
     """
 
-    time_range_constraint: Optional[TimeRangeConstraint]
+    _PREDICATE_METADATA_KEY = "is_predicate_property"
+
+    time_range_constraint: Optional[TimeRangeConstraint] = dataclasses.field(metadata={_PREDICATE_METADATA_KEY: True})
+    pushdown_state: PredicatePushdownState = PredicatePushdownState.FULLY_ENABLED
+
+    def __post_init__(self) -> None:
+        """Validation to ensure pushdown properties are configured correctly.
+
+        In particular, this asserts that cases where pushdown is disabled cannot leak pushdown operations via
+        outside property access - if pushdown is disabled, no further pushdown operations of any kind are allowed
+        on that particular code branch.
+        """
+        if self.pushdown_state is PredicatePushdownState.FULLY_ENABLED:
+            return
+
+        invalid_predicate_property_names = {
```

Comment:
So is this essentially just checking that `time_range_constraint` is always `None` when `pushdown_state` is `DISABLED`?
If so, this seems pretty complicated & hard to read for just that functionality, and it might be helpful to simplify by just checking that one scenario. But maybe I'm missing something or this gets used more generically down the stack?

Reply:
Oh yeah, I definitely should've explained that in the commit message. There will be other parameters added which will also need to be nulled out for disabled. Will update things accordingly.

I'm essentially trying to enforce that when the state is disabled there's no possibility of somebody doing something like `if pushdown_params.time_range_constraint or pushdown_params.where_filter_specs` and then triggering a pushdown operation. If the disabled state was binary I could just make it a derived property but because of our conversion metric and time window handling I have to deal with things like time range only and categorical only pushdown states.

Reply:
Sounds good!

Reply:
Sounds good!